src/sage/sets/disjoint_union_enumerated_sets.py

```diff
@@ -392,32 +392,145 @@ def __iter__(self):
+            sage: list(islice(iter(U), 4))
+            [(0, 1), (0, 2), (1, 1), (1, 2)]
+
+        Coverage test in case some element sets does not know it is finite
+        but correctly raises :exc:`StopIteration` during iteration::
+
+            sage: from sage.sets.set import Set_object
+            sage: class UnknowinglyFiniteSet(Set_object):
+            ....:     def is_finite(self):
+            ....:         return False
+            sage: [*UnknowinglyFiniteSet([1, 2, 3])]
+            [1, 2, 3]
+            sage: [*iter(UnknowinglyFiniteSet([1, 2, 3]))]
+            [1, 2, 3]
+            sage: list(islice(DisjointUnionEnumeratedSets(
+            ....:     (UnknowinglyFiniteSet(frozenset([1,2,3])), UnknowinglyFiniteSet(frozenset([4,5,6])))), 7))
+            [1, 2, 4, 3, 5, 6]
         """
-        for k in self._family.keys():
-            for el in self._family[k]:
+        def wrap_element(el, k):
+            nonlocal self
+            if self._keepkey:
+                el = (k, el)
+            if self._facade:
+                return el
+            else:
+                return self.element_class(self, el)  # Bypass correctness tests
+
+        keys_iter = iter(self._family.keys())
+        if self._keepkey:
+            seen_keys = []
+        el_iters = []
+        while keys_iter is not None or el_iters:
+            if keys_iter is not None:
+                try:
+                    k = next(keys_iter)
+                except StopIteration:
+                    keys_iter = None
+                if keys_iter is not None:
+                    el_set = self._family[k]
+                    try:
+                        is_finite = el_set.is_finite()
+                    except (AttributeError, TypeError, NotImplementedError):
+                        is_finite = False
+                    if is_finite:
+                        for el in el_set:
+                            yield wrap_element(el, k)
+                    else:
+                        el_iters.append(iter(el_set))
+                        if self._keepkey:
+                            seen_keys.append(k)
+            any_stopped = False
+            for i, obj in enumerate(zip(seen_keys, el_iters) if self._keepkey else el_iters):
+                if self._keepkey:
+                    k, el_iter = obj
+                else:
+                    k = None
+                    el_iter = obj
+                try:
+                    el = next(el_iter)
+                except StopIteration:
+                    el_iters[i] = None
+                    any_stopped = True
+                    continue
+                yield wrap_element(el, k)
+            if any_stopped:
                 if self._keepkey:
-                    el = (k, el)
-                if self._facade:
-                    yield el
+                    filtered = [*zip(
+                        *[(k, el_iter) for k, el_iter in zip(seen_keys, el_iters) if el_iter is not None])]
```

Comment:
```suggestion
                    filtered = list(zip(
                        *[(k, el_iter) for k, el_iter in zip(seen_keys, el_iters) if el_iter is not None]))
```
Personal preference, but I find the explicitness of this easier to parse.

Reply:
actually `list()` is slightly slower than `[*...]` since you need to lookup the global namespace, but the difference is negligible.

Reply:
Ah, yes. Anyways, thank you!